src/main/scala/nl.knaw.dans.easy.deposit/docs/DatasetXml.scala

```diff
@@ -0,0 +1,171 @@
+
+  private def authorDetails(author: Author, lang: Option[Attribute]) = {
+    if (author.surname.isEmpty)
+      author.organization.toSeq.map(orgDetails(author.role, lang))
+    else
+      <dcx-dai:author>
+        { optionalElem(author.titles, "dcx-dai:titles").addAttr(lang) }
+        { requiredElems(author.initials.map(Seq(_)), "dcx-dai:initials") }
+        { optionalElem(author.insertions, "dcx-dai:insertions") }
+        { requiredElems(author.surname.map(Seq(_)), "dcx-dai:surname") }
+        { optionalElem(author.role, "dcx-dai:role") }
+        { author.organization.toSeq.map(orgDetails(role = None, lang)) }
+      </dcx-dai:author>
+  }
+
+  private def orgDetails(role: Option[SchemedKeyValue[String]],
+                         lang: Option[Attribute])(organization: String) =
+      <dcx-dai:organization>
+        { role.toSeq.map(elem("dcx-dai:role", lang)) }
+        { requiredElems(Some(Seq(organization)),"dcx-dai:name").addAttr(lang) }
+      </dcx-dai:organization>
+
+  /** @param elem XML element to be adjusted */
+  implicit class RichElem(elem: Elem) extends Object {
+    // TODO make private once the thrown error can be tested through apply (when using targetFromQualifier for a non-enum)
+
+    /**
+     * @param target the default tag (namespace:label)
+     * @param source may have a qualifier or some other property that determines the tag
+     */
+    @throws(classOf[InvalidDocumentException])
+    def setTag[T](target: String, source: T): Elem = {
+      (source match {
+        case x: QualifiedSchemedValue[_, _] if target == targetFromQualifier => x.qualifier.toString
+        case a: Author if isRightsHolder(a) => "dcterms:rightsHolder"
+        case _ => target
+      }).split(":") match {
+        case Array(label) => elem.copy(label = label)
+        case Array(prefix, label) => elem.copy(prefix = prefix, label = label)
+        case a => throw InvalidDocumentException("DatasetMetadata", new Exception(
+          s"expecting (label) or (prefix:label); got [${ a.mkString(":") }] to adjust the <key> of ${ Utility.trim(elem) } created from: $source"
+        ))
+      }
+    }
+
+    def addAttr(maybeAttribute: Option[Attribute]): Elem = maybeAttribute.map(elem % _).getOrElse(elem)
+  }
+
+  /** @param elems the sequence of elements to adjust */
+  private implicit class RichElems(elems: Seq[Elem]) extends Object {
+    def addAttr(lang: Option[Attribute]): Seq[Elem] = elems.map(_.addAttr(lang))
+  }
+
+  private def requiredElems[T](source: Option[Seq[T]], target: String, lang: Option[Attribute] = None): Seq[Elem] = {
+    source.map(keepNonEmpty).filterNot(_.isEmpty).getOrElse{
+      throw InvalidDocumentException("DatasetMetadata",  new Exception(
+        s"no content for mandatory $target"
+      ))
+    }.map(elem(target, lang))
+  }
+
+  private def isRightsHolder[T](a: T with Author) = {
+    a.role.exists(_.key == "RightsHolder")
+  }
+
+  private def keepNonEmpty[T](ts: Seq[T]) = {
```

Comment:
shouldn't this also filter out the `Option.empty` cases?

Reply:
I got an headache figuring out the line that calls this. At that line I want a `None` when I have an `Option[Seq[String]]` with just empty strings, otherwise the empty strings weeded out of the `Seq`.

Now that I'm trying to explain, the `elems` method should also weed out the empty strings, and the generic classes in DatasetMetadata should all have a `require` that rejects mandatory fields with empty strings.

Reply:
May I suggest that (given that you got a headache from figuring out just a single line in this class) this is a sign that the `DatasetXml` class is probably over-engineered. Although you found a nice pattern, I think it is not very maintainable in this state, unless it will be simplified further. Given that you're currently doing just the 'simple' elements, I'm starting to worry how much effort it's going to cost you and/or other developers to add more complex elements in the future. Perhaps you should first focus on getting the XML generated in a simple way and refactor it later on. As you mentioned to me many times before:

> Premature optimization is the root of all evil.

![image](https://user-images.githubusercontent.com/5938204/41763818-8f979502-75ff-11e8-805d-b183e2701cea.png)